### PR TITLE
Add FilterController for slice-and-dice of a table of data

### PR DESCRIPTION
**Pull request by shepheb**

### CODE REVIEW

File: js/foam/u2/md/MDViewOverridesTrait.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+      this.Y.registerModel(this.Checkbox, 'foam.u2.tag.Checkbox');
+      this.Y.registerModel(this.TextField, 'foam.u2.TextField');
+      this.Y.registerModel(this.MultiLineTextField, 'foam.u2.MultiLineTextField');
+      this.Y.registerModel(this.Select, 'foam.u2.tag.Select');
+      this.SUPER();
+    },
+  ]
+});
+/**
+ * @fileoverview Description of this file.
```
Comment (adamvy) [inline js/foam/u2/md/MDViewOverridesTrait.js]: remove this

File: js/foam/u2/search/FilterController.js [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+CLASS({
+  package: 'foam.u2.search',
+  name: 'FilterController',
+  extends: 'foam.u2.View',
+
+  traits: [
+    'foam.u2.md.MDViewOverridesTrait',
```
Comment (adamvy) [inline js/foam/u2/search/FilterController.js]: Why did you need the trait rather than just extending DetailView ?

File: js/foam/u2/search/FilterController.js [hunk 3]
```diff
@@ -0,0 +1,451 @@
...
+    {
+      model_: 'foam.core.types.DAOProperty',
+      name: 'filteredDAO',
+      required: true,
+      toPropertyE: function(X) {
+        return X.lookup('foam.u2.md.TableView').create({
+          editColumnsEnabled: true,
+        }, X);
+      }
+      //REMOVE
```
Comment (adamvy) [inline js/foam/u2/search/FilterController.js]: Actually remove this.

File: js/foam/u2/search/FilterController.js [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+      postSet: function(old, nu) {
+        if (nu) {
+          nu.select(COUNT())(function(c) {
+            this.totalCount = c.count;
+          }.bind(this));
+        }
+      }
+    },
+    {
+      model_: 'foam.core.types.DAOProperty',
```
Comment (adamvy) [inline js/foam/u2/search/FilterController.js]: Switch to type: ''

File: js/foam/u2/search/FilterController.js [hunk 5]
```diff
@@ -0,0 +1,451 @@
...
+      name: 'searchMgr',
+      lazyFactory: function() {
+        return this.SearchMgr.create({
+          dao$: this.data$,
+          filteredDAO$: this.filteredDAO$,
+        });
+      }
+    },
+    {
+      model_: 'StringArrayProperty',
```
Comment (adamvy) [inline js/foam/u2/search/FilterController.js]: ditto

File: js/foam/u2/search/FilterController.js [hunk 6]
```diff
@@ -0,0 +1,451 @@
...
+                  this.GroupBySearchView : this.TextSearchView;
+              var options = {
+                floatingLabel: false,
+                name: nu[i]
+              };
+              if (prop.tableSeparator) {
+                options.split = prop.tableSeparator;
+              }
+              this.addGroup(model, prop, options);
+              this.renderFilter(nu[i]);
```
Comment (adamvy) [inline js/foam/u2/search/FilterController.js]: Seems like a race condition.

We should guard this on this.filtersE_ being set, or put a factory on this.filtersE_ so that we don't die if this postSet happens before initE.

File: js/foam/u2/search/GroupBySearchView.js [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+      defaultValue: true
+    }
+  ],
+
+  methods: [
+    function clear() {
+      this.view.data = '';
+    },
+    function initE() {
+      if (this.floatingLabel) return this.view.initE.call(this);
```
Comment (adamvy) [inline js/foam/u2/search/GroupBySearchView.js]: This seems wrong, why not just this.add(this.view); return;

File: js/foam/u2/search/FilterController.js (hunk 2 above)
Comment (shepheb) [inline js/foam/u2/search/FilterController.js]: I'm under the impression that choosing which `View` type to extend was a Bad Thing about U1 that we're glad to be free of, and that `u2.DetailView` is a leaf, not the root of a large subtree.

To that end, I moved this logic into a trait so it could be reused by others who want these overrides, including users' own controllers.

File: js/foam/u2/search/GroupBySearchView.js (hunk 7 above)
Comment (shepheb) [inline js/foam/u2/search/GroupBySearchView.js]: That's better, I think, and it should work.

File: js/foam/u2/search/FilterController.js (hunk 6 above)
Comment (shepheb) [inline js/foam/u2/search/FilterController.js]: Easy to turn filtersE_ into a factory that's added at the right point in the template. It's just an empty `div` with a class.

File: js/foam/u2/search/FilterController.js (hunk 4 above)
Comment (shepheb) [inline js/foam/u2/search/FilterController.js]: Done.

File: js/foam/u2/search/FilterController.js (hunk 5 above)
Comment (shepheb) [inline js/foam/u2/search/FilterController.js]: Done.

File: js/foam/u2/search/FilterController.js (hunk 3 above)
Comment (shepheb) [inline js/foam/u2/search/FilterController.js]: Done.

File: js/foam/u2/md/MDViewOverridesTrait.js (hunk 1 above)
Comment (shepheb) [inline js/foam/u2/md/MDViewOverridesTrait.js]: Whoops. Removed.

**Comment by shepheb** [comment]
All addressed. Only remaining open question is why the trait rather than extending `DetailView`.

File: js/foam/u2/search/FilterController.js (hunk 6 above)
Comment (shepheb) [inline js/foam/u2/search/FilterController.js]: Which is to say, done.

File: js/foam/u2/search/TextSearchView.js [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+    {
+      name: 'name',
+      documentation: 'All SearchViews require a name. Defaults to "query".',
+      defaultValue: 'query',
+    },
+  ],
+
+  methods: [
+    function initE() {
+      this.cls('foam-u2-search-filter').cls(this.myCls());
```
Comment (adamvy) [inline js/foam/u2/search/TextSearchView.js]: Where is foam-u2-search-filter defined?

Seems flaky to depend upon CSS you don't define nor inherit.
